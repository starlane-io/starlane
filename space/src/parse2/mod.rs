mod chars;
mod primitive;
mod scaffold;
mod token;
mod err;

use std::error::Error;
use crate::parse::util::{preceded, Span};
use ariadne::{Label, Report, ReportKind, Source};
use nom::character::complete::alpha1;
use nom::combinator::all_consuming;
use nom::error::{
    ErrorKind, FromExternalError, ParseError,
};
use nom::multi::separated_list1;
use nom::sequence::pair;
use nom::{Compare, Finish, IResult, InputLength, InputTake, Offset, Parser};
use nom_locate::LocatedSpan;
use nom_supreme::context::ContextError;
use nom_supreme::parser_ext::ParserExt;
use std::fmt::{Debug, Display, Formatter};
use std::ops::Range;
use nom::bytes::complete::tag;
use strum_macros::{Display, EnumString};
use nom_supreme::final_parser::ExtractContext;
use nom_supreme::tag::TagError;
use thiserror::Error;
use nom_supreme::error::{ErrorTree, BaseErrorKind, Expectation, GenericErrorTree};
type Input<'a> = LocatedSpan<&'a str, ParseOpRef<'a>>;

pub fn range(input: &Input) -> Range<usize> {
    input.location_offset()..input.fragment().len() 
}


/// `op` is a helpful name of this parse operation i.e. `BindConf`,`PackConf` ...
pub fn parse_operation(op: impl ToString, data: &str) -> ParseOp {
    ParseOp::new(op.to_string(), data)
}


pub type ParseErrs<'a> = GenericErrorTree<Input<'a>, &'static str, Ctx, Box<dyn Error + Send + Sync + 'static>>;
pub type Res<'a, O> = IResult<Input<'a>, O, ParseErrs<'a>>;

pub trait Operation {}
pub struct ParseOperationDef<'a, N, S> {
    name: N,
    stack: S,
    data: &'a str,
}

impl<'a, N, S> ParseOperationDef<'a, N, S> {
    fn from(name: N, stack: S, data: &'a str) -> Self {
        Self { name, stack, data }
    }
    pub fn data(&self) -> &'a str {
        self.data
    }
}

impl<'a> Debug for ParseOpRef<'a> {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "ParseOp: '{}' data.len(): {}  ",
            self.name,
            self.data.len()
        )?;
        Ok(())
    }
}

pub type ParseOp<'a> = ParseOperationDef<'a, String, Vec<Ctx>>;

impl<'a> ParseOp<'a> {
    fn new(name: String, data: &'a str) -> Self {
        Self::from(name, Default::default(), data)
    }
    fn ctx_range(&self, range: Range<usize>) -> &[Ctx] {
        let blah = self.stack.as_slice();
        &blah[range]
    }

    fn push(&mut self, ctx: Ctx) {
        self.stack.push(ctx);
    }

    fn to_ref(&'a self) -> ParseOpRef<'a> {
        let name: &'a str = self.name.as_str();
        let stack: &'a [Ctx] = self.stack.as_slice();
        ParseOpRef::from(name, stack, self.data)
    }

    fn input(&'a self) -> Input<'a> {
        Input::new_extra(self.data, self.to_ref())
    }
}

pub type ParseOpRef<'a> = ParseOperationDef<'a, &'a str, &'a [Ctx]>;
impl<'a> ParseOpRef<'a> {
    /// returning [Self] from [Range] instead of [Result<Self,()>]
    /// goes against `rust's` principles, however, since this `unsafe`
    /// this code is only used by this parsing mod and I think it will
    /// become robust over time.  Since the [ParseOpRef] is integral
    /// in managing parse errors it's a little hard to do proper error
    /// [Result] on the error system!
    fn slice(& self, range: Range<usize>) -> Self {
        let stack = & self.stack[range];
        Self {
            name: self.name,
            stack,
            data: self.data,
        }
    }
}

impl<'a> Clone for ParseOpRef<'a> {
    fn clone(&self) -> Self {
        Self::from(self.name, self.stack, self.data)
    }
}

struct ParseErrsFinal {
    pub string: String,
    pub errors: Vec<(Range<usize>, ErrKind)>,
}

/*
struct ParseErrs {
    pub errors: Vec<(Range<usize>, ErrKind)>,
}

impl Default for ParseErrs {
    fn default() -> Self {
        Self { errors: Default::default() }
    }
}

impl ParseErrs {
    pub fn panic(m: impl ToString) -> Self {
        Self { errors: vec![(0..1,ErrKind::Panic(m.to_string()))] }
    }
}

impl Debug for ParseErrs {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        for (input, error) in &self.errors {
            write!(f, "{} in {}..{}", error, input.start, input.end)?;
        }
        Ok(())
    }
}

 */

#[derive(Debug, Eq, PartialEq, Hash, Clone, Error)]
pub enum ErrKind {
    #[error("Tag({0})")]
    Tag(&'static str),
    #[error("ErrKind")]
    Nom(ErrorKind),
    #[error("not expecting: '{0}'")]
    Char(char),
    #[error("{0}")]
    Context(Ctx),
    #[error("{0}")]
    Panic(String),
}

pub struct ErrorKindWrap(ErrKind);

impl Display for ErrorKindWrap {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.0.to_string())
    }
}

#[derive(Debug, Eq, PartialEq, Hash, Clone, EnumString, Display)]
pub enum Ctx {
    Token,
    #[strum(to_string="upper case alphabetic character")]
    UpperCaseChar,
    #[strum(to_string="lower case alphabetic character")]
    LowerCaseChar,   
    #[strum(to_string="Camel Case")]
    CamelCase,
    #[strum(to_string="Skewer Case")]
    SkewerCase,
    #[strum(to_string="Snake Case")]
    SnakeCase,
    #[strum(to_string="Type")]
    Type,
    #[strum(to_string="Class")]
    Class,
    #[strum(to_string="Data")]
    Data 
}

/*
impl<'a> ParseError<Input<'a>> for ParseErrs {
    fn from_error_kind(input: Input<'a>, kind: ErrorKind) -> Self {
        Self {
            errors: vec![(range(input), ErrKind::Nom(kind))],
        }
    }

    fn append(input: Input<'a>, kind: ErrorKind, mut other: Self) -> Self {
        other.errors.push((range(input), ErrKind::Nom(kind)));
        other
    }

    fn from_char(input: Input<'a>, c: char) -> Self {
        Self {
            errors: vec![(range(input), ErrKind::Char(c))],
        }
    }
}

 */
/*
impl<'a> ContextError<Input<'a>, Ctx> for ParseErrs {
    fn add_context(input: Input<'a>, ctx: Ctx, mut other: Self) -> Self {
        other.errors.push((range(input), ErrKind::Context(ctx)));
        other
    }
}

impl <'a> TagError<Input<'a>, &'static str> for ParseErrs {
    fn from_tag(input: Input, tag: &'static str) -> Self {
        let mut errs = ParseErrs::default();
        errs.errors.push((range(input), ErrKind::Tag(tag)));
        errs
    }
}

 */

/*
fn segments(ix : Input) -> Res < Vec < Input > >
{
    let mut f = move | ix2| 
        {
            let mut parser =
                pair(separated_list1(tag(":"), alpha1 :: < Input, ParseErrs >),
                     preceded(tag("^").context(Ctx :: Yuk), alpha1));
            
            parser.parse(ix2).map(| (next, (segments, extra)) | (next, segments))
        };
    
    expect(f , "segment", "x", "y") (ix)
}

 */


pub fn segments(i: Input) -> Res<Vec<Input>> {
    let mut parser = pair(
        separated_list1(tag(":"), alpha1::<Input, ParseErrs>),
        preceded(tag("^"), alpha1),
    );

    parser
        .parse(i)
        .map(|(next, (segments, extra))| (next, segments))
}

pub fn camel(input: Input) -> Res<String> {
    chars::camel(input).map(|(next,rtn)|(next, rtn.to_string()))
}


/*
#[push_ctx_for_input]

pub fn segments(i: Input) -> Res<Vec<Input>> {
    let mut parser = pair(
        separated_list1(tag(":"), alpha1::<Input, ParseErrs>),
        preceded(tag("^").context(Ctx::Yuk), alpha1),
    );

    parser
        .parse(i)
        .map(|(next, (segments, extra))| (next, segments))
}

 */



#[test]
fn test() {
    let op = parse_operation("test", "you:are:^awesome");
    let input = op.input();
    let error = all_consuming(segments)(input).finish().unwrap_err();
    log(op.data,error);
}

fn log(data: impl AsRef<str>, err: ParseErrs) {
    match &err {
        ParseErrs::Base { location, ref kind } => {
            
            let range = range(&location);

            let mut builder = Report::build(ReportKind::Error, range.clone());
            match kind {
                BaseErrorKind::Expected(expect) => {
                    let report = builder.with_message(format!("Expected: '{}' found: {}", expect, location)).with_label(
                        Label::new(range)
                            .with_message(format!("{}",err)),
                    ).finish();
                    report.print(Source::from(data.as_ref())).ok();
                }
                BaseErrorKind::Kind(kind) => {}
                BaseErrorKind::External(external) => {}
            }
        }
        ParseErrs::Stack { base, contexts } => {
            panic!();
        }
        ParseErrs::Alt(_) => {

            panic!();
        }
    }
    todo!()
    /*
    for (range, err) in err.errors {
        Report::build(ReportKind::Error, range.clone())
            .with_message(err.to_string())
            .with_label(
                Label::new(range)
                    .with_message(format!("{}",err)),
            )
            .finish()
            .print(Source::from(data.as_ref()))
            .unwrap();
    }
    
     */
}
