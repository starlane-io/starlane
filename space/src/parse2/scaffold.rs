use crate::types2::archetype::Archetype;

pub struct PairScaffold<Prefix, Infix>
where
    Prefix: Archetype,
{
    prefix: Prefix,
    infix: Infix,
}
