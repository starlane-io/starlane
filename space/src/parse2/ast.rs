use std::fmt::Debug;
use std::marker::PhantomData;
use std::sync::Arc;
use indexmap::IndexMap;
use crate::parse2::token::{TokenIter, DocType, Token, TokenKind, Tokens};
use semver::Version;
use strum_macros::{Display, EnumString};
use crate::parse2::ast::err::AstErr;
use crate::parse2::ast::package::header_decl;
use crate::parse2::document::{Declarations, Definitions, DocumentDef, Unit};
use crate::parse2::err::{ParseErrs2, ParseErrs2Def, ParseErrs2Proto};
use crate::parse2::{Input, ParseResultProto};
use crate::parse::model::{BlockSymbol, LexBlock, NestedSymbols};

pub(crate) fn ast<'a>(tokens: Tokens<'a>) -> ParseResultProto<'a> {
    let errs = ParseErrs2Def::new();
    let mut iter = tokens.iter();
    todo!();
}


pub enum Ast<'a> {
    Definitions(Block<'a,Definitions<'a>>),
    Err(ParseErrs2<'a>)
}

mod package {
    use crate::parse2::ast::err::AstErr;
    use crate::parse2::ast::Header;
    use crate::parse2::err::{ParseErrs2Def, ParseErrs2Proto};
    use crate::parse2::token::{TokenIter, DocType, Ident, TokenKind};

    pub fn header_decl<'a>(iter: &'a mut TokenIter<'a>) -> Result<Header, ParseErrs2Proto<'a>> {
        /// if it succe
        iter.expect("Document Type",&TokenKind::Ident(Ident::Camel(DocType::Package.into())))?;
        let doc_type = DocType::Package;
        todo!();
    }
}



#[derive(Debug,Clone)]
struct Header {
    doc_type: DocType,
    version: Version
}

pub mod err {
    use crate::parse::CamelCase;
    use crate::parse2::token::{Token, TokenKind, TokenKindDisc};
    use crate::parse2::{range, Input, Op};
    use ariadne::{Label, Report, ReportKind, Source};
    use std::fmt::{Display, Formatter};
    use std::ops::{Deref, DerefMut};
    use thiserror::Error;
    use crate::parse2::document::Unit;
    use crate::parse2::err::{ParseErrs2Def, ParseErrs2Proto};

    #[derive(Clone)]
    pub struct Errs<'a> {
        errs: Vec<AstErr<'a>>,
    }
    
    impl <'a> Default for Errs<'a> {
        fn default() -> Self {
            Self::new()
        }
    }
    impl <'a> Errs<'a> {
        
        pub fn new() -> Self {
            Self {
                errs: Default::default()
            }
        } 
    }
    
    impl <'a> Deref for Errs<'a> {
        type Target = Vec<AstErr<'a>>;

        fn deref(&self) -> &Self::Target {
            & self.errs
        }
    }
    
    impl <'a> DerefMut for Errs<'a> {
        fn deref_mut(&mut self) -> &mut Self::Target {
           & mut self.errs
        }
    }
        

    pub type AstErr<'a> = Unit<'a,AstErrKind>;
   
    impl <'a> Into<ParseErrs2Proto<'a>> for AstErr<'a> {
        fn into(self) -> ParseErrs2Proto<'a> {
            let mut errs = ParseErrs2Proto::new();
            errs.add(self);
            errs
        }
    }

    impl<'a> Display for AstErr<'a> {
        fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
                write!(f, "{err} -- range: [{}..{}]: {}", self.span.location_offset(), (self.span.location_offset() + self.span.fragment().len()),self.kind)
        }
    }

    #[derive(Debug, Clone, Error)]
    pub enum AstErrKind {
        #[error("Document type not recognized: '{0}'")]
        DocumentTypeNotRecognized(CamelCase),
        #[error("Expected: {id} {kind}' found: '{found}'")]
        ExpectedKind { id: &'static str, kind: TokenKind, found: TokenKind },
        #[error("Expected '{literal}' found: '{found}'")]
        ExpectedLiteral{ literal: &'static str, found: TokenKind },
        #[error("Expected token type: '{0}' but instead reach EOF (End of File)")]
        UnexpectedEof(TokenKind),
        #[error("Expected whitespace. Found: '{0}'")]
        Whitespace(TokenKind),
        #[error("Version Format Error. Expecting format: 'major.minor.patch-release+label' i.e.: '3.7.8', '2.0.5-rc', '1.2.1-beta+preview'")]
        VersionFormat,
    }

    impl AstErrKind {
        pub fn with(self, span: Input) -> AstErr {
            AstErr::new(span,self)
        }
    }
    
    impl ParseErrs2Def {
        pub fn report(&self, errs: &Vec<AstErr>) {
            let r = 0..self.data.len();
            let mut builder = Report::build(ReportKind::Error, r.clone());
            //for err in errs {
            let err = errs.first().unwrap();
            match err {
                AstErr::Token { token, err } => {
                    let report = builder
                        .with_message("some errors")
                        .with_label(Label::new(range(&token.span)).with_message(err.to_string()))
                        .finish();

                    report.print(Source::from(self.data.as_str())).ok();
                }
                AstErr::Err(_) => {
                    panic!();
                }
                // }
            }
        }
    }
}


pub struct Alt<P,O> {
    branch: Vec<Branch<P,O>>
}

impl <P,O> Alt<P,O> {
    pub fn add( & mut self, alt: Branch<P,O> ) {
        self.branch.push(alt)
    }
}

struct Branch<P,O> where P: AstParser, O: Into<Ast> {
    /// preparser peeks ahead for pattern
    pub pre: P,
    /// inside the block 
    pub block: Box<dyn AstParser<Output=O>>,
}

impl <P,O> Branch<P,O> where P: AstParser, O: Into<Ast> {
    pub fn new(pre: P, block: impl AstParser<Output=O>) -> Self {
        Self {
            pre,
            block: Box::new(block)
        }
    }
}


pub trait AstParser<'a> {
  type Output: Into<Ast<'a>>;
  fn parse(&self, tokens: &'a mut TokenIter<'a>) -> Result<Self::Output,ParseErrs2Proto<'a>>;
}


#[derive(Debug,Clone)]
pub enum AstBlock<'a> {
    /// `Defs`(version=1.0.1)  { ... }
    Header(Header),
    /// + `arg` {  ... }
    Arg(Declarations<'a>),
    /// + `env` { ... }
    Env(Declarations<'a>),
    /// + `properties`
    PropertyDefs,
    /// a line block is terminated by a semi-colon `;`
    /// `some-param[str] = something`;
    Line,
}


#[derive(Debug,Clone)]
struct Block<'a,C> where C: 'a+Debug+Clone {
    kind: AstBlock<'a>,
    symbol: BlockSymbol,
    open: Unit<'a,&'static str>,
    close: Unit<'a,&'static str>,
    content: Unit<'a,C>
}

pub struct DelimitedParser<'a,O> {
    pre: Box<dyn AstParser<Output=_>>,
    content: Box<dyn AstParser<Output=O>>,
    post: Box<dyn AstParser<Output=_>>,
}

struct LiteralParser(TokenKind);

impl LiteralParser {
    pub fn new(kind: TokenKind) -> Self {
        Self(kind)
    }
}

impl <'a> AstParser<'a> for LiteralParser {
    type Output = &'a Token<'a>;

    fn parse(&self, tokens: &'a mut TokenIter<'a>) -> Result<Self::Output, ParseErrs2Proto<'a>> {
        if let Some(token) = tokens.expect("literal", self.0.clone().into() ) && self.0 == token.kind {
            Ok(token)
        } else {
            let mut errs = ParseErrs2Proto::new();
            let err = AstErr::
            errs.add( )
            
        }
    }
}



impl <'a,O> DelimitedParser<'a,O> {
    pub fn new(kind: BlockSymbol) -> Self {
        
    }  
}



impl <'a,I> AstParser for BlockParser<'a,I> where I: AstParser {
    type Output = Block<'a,I>;
    fn parse<'a>(tokens: &'a mut TokenIter<'a>) -> Result<Self::Output, ParseErrs2Proto<'a>> {
        
    }
}


impl <'a,P> BlockParser<P> where P: AstParser<Output=Ast<'a>> {
    pub fn new(kind: BlockSymbol) -> Self {
        LexBlock
        Self {
            kind,
            inner: PhantomData::default()
        }
    } 
}




