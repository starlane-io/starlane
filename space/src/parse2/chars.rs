use crate::parse::util::recognize;
use crate::parse::{CamelCase, SkewerCase, SnakeCase};
use crate::parse2::chars::recognize::{
    lower1, lower_alphanumeric_plus_dash0, lower_alphanumeric_plus_underscore0, upper1,
};
use crate::parse2::token::Ident;
use crate::parse2::{Ctx, Input, Res};
use nom::branch::alt;
use nom::character::complete::alphanumeric0;
use nom::character::streaming::alphanumeric1;
use nom::combinator::{into, value};
use nom::multi::many1;
use nom::sequence::pair;

pub fn camel(input: Input) -> Res<CamelCase> {
    recognize::camel(input).map(|(next, rtn)| (next, CamelCase::new(rtn.to_string())))
}

pub fn skewer(input: Input) -> Res<SkewerCase> {
    recognize::skewer(input).map(|(next, rtn)| (next, SkewerCase::new(rtn.to_string())))
}

pub fn snake(input: Input) -> Res<SnakeCase> {
    recognize::camel(input).map(|(next, rtn)| (next, SnakeCase::new(rtn.to_string())))
}

fn undefined(input: Input) -> Res<Input> {
    use recognize::*;
    recognize(many1(alt((alphanumeric1, dash, underscore))))(input)
}

pub fn ident(input: Input) -> Res<Ident> {
    alt((into(camel), into(skewer), into(snake)))(input)
}

mod recognize {
    use crate::parse::util::recognize;
    use crate::parse2::{Ctx, Input, Res};
    use nom::branch::alt;
    use nom::bytes::complete::{is_a, tag};
    use nom::character::complete::{alpha1, alphanumeric0, alphanumeric1};
    use nom::combinator::opt;
    use nom::sequence::pair;
    use nom_supreme::ParserExt;

    pub fn dash(input: Input) -> Res<Input> {
        recognize(tag("-"))(input)
    }

    pub fn newline(input: Input) -> Res<Input> {
        recognize(tag("\n"))(input)
    }

    pub fn underscore(input: Input) -> Res<Input> {
        recognize(tag("_"))(input)
    }

    pub fn upper1(input: Input) -> Res<Input> {
        recognize(is_a("ABCDEFGHIJKLMNOPQRSTUVWXYZ"))(input)
    }

    pub fn lower1(input: Input) -> Res<Input> {
        recognize(is_a("abcdefghijklmnopqrstuvwxyz"))(input)
    }

    pub fn lower_alphanumeric_plus_dash0(input: Input) -> Res<Input> {
        recognize(opt(alt((lower1, tag("-")))))(input)
    }

    pub fn lower_alphanumeric_plus_dash1(input: Input) -> Res<Input> {
        recognize(alt((lower1, tag("-"))))(input)
    }

    pub fn lower_alphanumeric_plus_underscore0(input: Input) -> Res<Input> {
        recognize(opt(alt((lower1, tag("_")))))(input)
    }

    pub fn lower_alphanumeric_plus_underscore1(input: Input) -> Res<Input> {
        recognize(alt((lower1, tag("_"))))(input)
    }

    pub fn camel(input: Input) -> Res<Input> {
        recognize(pair(upper1, alphanumeric0).context(Ctx::CamelCase))(input)
    }

    pub fn skewer(input: Input) -> Res<Input> {
        recognize(pair(lower1, lower_alphanumeric_plus_dash0).context(Ctx::SkewerCase))(input)
    }

    pub fn snake(input: Input) -> Res<Input> {
        recognize(pair(lower1, lower_alphanumeric_plus_underscore0).context(Ctx::SkewerCase))(input)
    }
}
