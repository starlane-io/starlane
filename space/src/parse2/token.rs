use crate::parse::model::NestedBlockKind;
use crate::parse::util::Span;
use crate::parse::{CamelCase, Domain, SkewerCase, SnakeCase};
use crate::parse2::chars::ident;
use crate::parse2::token::block::block;
use crate::parse2::token::punctuation::punctuation;
use crate::parse2::{ErrTree, Input, ParseErrsTree, Res};
use derive_builder::Builder;
use nom::branch::alt;
use nom::error::ParseError;
use nom::multi::many1;
use nom::{Offset, Parser, Slice};
use nom_supreme::ParserExt;
use semver::Version;
use std::collections::HashMap;
use std::ops::Range;
use std::str::FromStr;
use strum_macros::{Display, EnumDiscriminants};


fn token(input: Input) -> Res<Token> {
    use whitespace::whitespace;
    alt((tok(whitespace),tok(punctuation), tok(ident), tok(block)))(input)
}

fn tokenize(input: Input) -> Res<Vec<Token>> {
    many1(token)(input).map(|(next,tokens)|{
        println!("**********************");
        println!("tokens.len(): {}", tokens.len());
        for token in &tokens {
            println!("{}", token.token);
        }
        println!("**********************");
        
        (next,tokens)})
}


#[derive(Clone, Debug)]
pub struct Token {
    span: Range<usize>,
    token: TokenKindDef,
}

#[derive(Clone, Debug, EnumDiscriminants, Display)]
#[strum_discriminants(vis(pub))]
#[strum_discriminants(name(TokenKind))]
#[strum_discriminants(derive(Hash,Display))]
enum TokenKindDef {
    #[strum(to_string="Ident({0})")]
    Ident(Ident),
    #[strum(to_string="BlockOpen({0})")]
    Open(NestedBlockKind),
    #[strum(to_string="BlockClose({0})")]
    Close(NestedBlockKind),
    /// `+` symbol
    Plus,
    /// `@` symbol
    At,
    /// `:` symbol
    SegmentSep,
    /// `::` symbol
    Scope,
    /// `+::` add variant
    Variant,
    /// `.` symbol (used for properties and child defs)
    Dot,
    /// semicolon ';' (just as god intended)
    Terminator,
    /// `&` good old ampersand
    Return,
    /// `version=` i.e.: Def(`version=`1.1.5) ... tells which parser version to use
    VersionPrelude,
    /// a `space` or a `tab`
    Space,
    /// any cluster of whitespace: `space`, `tab` and `newline`
    Newline,
    /// an erroneous token...
    Err(Range<usize>),
}

impl TokenKindDef {
    fn open(block: NestedBlockKind) -> Self {
        Self::Open(block)
    }

    fn close(block: NestedBlockKind) -> Self {
        Self::Close(block)
    }
}
#[derive(Clone, Debug, EnumDiscriminants, Display)]
#[strum_discriminants(vis(pub))]
#[strum_discriminants(name(IdentKind))]
#[strum_discriminants(derive(Hash))]
pub(crate) enum Ident {
    #[strum(to_string="Camel({0})")]
    Camel(CamelCase),
    #[strum(to_string="Skewer({0})")]
    Skewer(SkewerCase),
    #[strum(to_string="Snake({0})")]
    Snake(SnakeCase),
    #[strum(to_string="Domain({0})")]
    Domain(Domain),
    #[strum(to_string="Version({0})")]
    Version(Version),
    #[strum(to_string="Undefined({0})")]
    /// [Ident::Undefined] represents a semi plausible ident ... maybe camel case with underscores & dashes
    Undefined(String),
}

impl From<Ident> for TokenKindDef {
    fn from(ident: Ident) -> Self {
        TokenKindDef::Ident(ident)
    }
}

impl<'a> From<Input<'a>> for Ident {
    fn from(value: Input<'a>) -> Self {
        Self::Undefined(value.to_string())
    }
}

impl From<CamelCase> for Ident {
    fn from(value: CamelCase) -> Self {
        Ident::Camel(value)
    }
}
impl From<SkewerCase> for Ident {
    fn from(value: SkewerCase) -> Self {
        Ident::Skewer(value)
    }
}
impl From<SnakeCase> for Ident {
    fn from(value: SnakeCase) -> Self {
        Ident::Snake(value)
    }
}
impl From<Domain> for Ident {
    fn from(value: Domain) -> Self {
        Ident::Domain(value)
    }
}
impl From<Version> for Ident {
    fn from(value: Version) -> Self {
        Ident::Version(value)
    }
}

#[derive(Clone, Builder)]
pub struct DocTokenized {
    pub kind: Token,
    pub version: Token,
    pub defs: HashMap<Ident, Vec<Token>>,
}

#[derive(Clone, Builder)]
pub struct DocLayer {
    pub kind: Token,
    pub defs: HashMap<Ident, Vec<Token>>,
}

pub(super) mod block {
    use crate::parse2::token::TokenKindDef;
    use crate::parse2::{Input, Res};
    use nom::branch::alt;

    pub fn block(input: Input) -> Res<TokenKindDef> {
        use close::token as close;
        use open::token as open;
        alt((open, close))(input)
    }

    pub(super) mod open {
        use crate::parse::model::NestedBlockKind;
        use crate::parse2::token::TokenKindDef;
        use crate::parse2::{Input, Res};
        use nom::branch::alt;
        use nom::Parser;
        use nom_supreme::tag::complete::tag;
        use nom_supreme::ParserExt;

        pub fn angle(input: Input) -> Res<NestedBlockKind> {
            tag("<")(input).map(|(next, _)| (next, NestedBlockKind::Angle))
        }

        pub fn square(input: Input) -> Res<NestedBlockKind> {
            tag("[")(input).map(|(next, _)| (next, NestedBlockKind::Square))
        }

        pub fn parenthesis(input: Input) -> Res<NestedBlockKind> {
            tag("(")(input).map(|(next, _)| (next, NestedBlockKind::Parens))
        }

        pub fn curly(input: Input) -> Res<NestedBlockKind> {
            tag("{")(input).map(|(next, _)| (next, NestedBlockKind::Curly))
        }

        pub fn open(input: Input) -> Res<NestedBlockKind> {
            alt((angle, square, parenthesis, curly))(input)
        }

        pub fn token(input: Input) -> Res<TokenKindDef> {
            open(input).map(|(next, kind)| {
                let kind = TokenKindDef::Open(kind);
                (next, kind)
            })
        }
    }

    pub(super) mod close {
        use crate::parse::model::NestedBlockKind;
        use crate::parse2::token::TokenKindDef;
        use crate::parse2::{Input, Res};
        use nom::branch::alt;
        use nom::Parser;
        use nom_supreme::tag::complete::tag;

        pub fn angle(input: Input) -> Res<NestedBlockKind> {
            tag(">")(input).map(|(next, _)| (next, NestedBlockKind::Angle))
        }

        pub fn square(input: Input) -> Res<NestedBlockKind> {
            tag("]")(input).map(|(next, _)| (next, NestedBlockKind::Square))
        }

        pub fn parenthesis(input: Input) -> Res<NestedBlockKind> {
            tag(")")(input).map(|(next, _)| (next, NestedBlockKind::Parens))
        }

        pub fn curly(input: Input) -> Res<NestedBlockKind> {
            tag("}")(input).map(|(next, _)| (next, NestedBlockKind::Curly))
        }

        pub fn close(input: Input) -> Res<NestedBlockKind> {
            alt((angle, square, parenthesis, curly))(input)
        }
        pub fn token(input: Input) -> Res<TokenKindDef> {
            close(input).map(|(next, kind)| {
                let kind = TokenKindDef::Close(kind);
                (next, kind)
            })
        }
    }
}

pub(super) mod whitespace {
    use crate::parse2::token::TokenKindDef;
    use crate::parse2::{Input, Res};
    use nom::branch::alt;
    use nom::bytes::complete::tag;
    use nom::character::complete::space1;
    use nom::combinator::value;
    use nom::multi::many0;

    pub fn space(input: Input) -> Res<TokenKindDef> {
        value(TokenKindDef::Space, space1)(input)
    }

    pub fn newline(input: Input) -> Res<TokenKindDef> {
        value(TokenKindDef::Newline, tag("\n"))(input)
    }
    
    pub fn whitespace(input: Input) -> Res<TokenKindDef> {
        alt((space,newline))(input)
    }

    pub fn multi0(input: Input) -> Res<Vec<TokenKindDef>> {
        many0(alt((space, newline)))(input)
    }

    pub fn multi1(input: Input) -> Res<Vec<TokenKindDef>> {
        many0(alt((space, newline)))(input)
    }
}

pub(super) mod punctuation {
    use crate::parse2::token::TokenKindDef;
    use crate::parse2::{Input, Res};
    use nom::branch::alt;
    use nom::bytes::complete::tag;
    use nom::combinator::value;

    fn at(input: Input) -> Res<TokenKindDef> {
        value(TokenKindDef::At, tag("@"))(input)
    }

    fn plus(input: Input) -> Res<TokenKindDef> {
        value(TokenKindDef::Plus, tag("+"))(input)
    }

    fn segment_sep(input: Input) -> Res<TokenKindDef> {
        value(TokenKindDef::SegmentSep, tag(":"))(input)
    }

    fn scope(input: Input) -> Res<TokenKindDef> {
        value(TokenKindDef::Scope, tag("::"))(input)
    }

    fn dot(input: Input) -> Res<TokenKindDef> {
        value(TokenKindDef::Dot, tag("."))(input)
    }

    fn terminator(input: Input) -> Res<TokenKindDef> {
        value(TokenKindDef::Terminator, tag(";"))(input)
    }

    fn r#return(input: Input) -> Res<TokenKindDef> {
        value(TokenKindDef::Return, tag("&"))(input)
    }

    pub(super) fn punctuation(input: Input) -> Res<TokenKindDef> {
        alt((scope, segment_sep, dot, plus, at, terminator, r#return))(input)
    }
}

fn kind<O>(mut f: impl FnMut(Input) -> Res<O>) -> impl FnMut(Input) -> Res<TokenKindDef>
where
    O: Into<TokenKindDef>,
{
    move |input| {
        let (next, output) = f(input)?;
        let kind = output.into();
        Ok((next, kind))
    }
}

fn tok<O>(mut f: impl FnMut(Input) -> Res<O>+Copy) -> impl FnMut(Input) -> Res<Token>
where
    O: Into<TokenKindDef>,
{
    move |input| {
        let (next, output) = kind(f)(input.clone())?;
        let len = (next.location_offset() - input.location_offset());
        let span = input.location_offset()..(input.location_offset() + len);
        let kind = output.into();
        let token = Token { span, token: kind };
        Ok((next, token))
    }
}

pub mod err {
    use crate::parse2::Input;
    use strum_macros::Display;
    use thiserror::Error;

    #[derive(Clone, Display, Debug, Error)]
    pub enum TokenErr<'a> {
        Expect {
            expected: &'static str,
            found: Input<'a>,
        },
    }

    impl<'a> TokenErr<'a> {
        pub fn expected(expected: &'static str, found: Input<'a>) -> Self {
            Self::Expect { expected, found }
        }
    }
}

pub fn result<R>(result: Res<R>) -> Result<R, ErrTree> {
    match result {
        Ok((_, e)) => Ok(e),
        Err(nom::Err::Error(err)) => Result::Err(err),
        Err(nom::Err::Failure(err)) => Result::Err(err),
        Err(nom::Err::Incomplete(_)) => panic!("nom::Err::Incomplete not implemented "),
    }
}

#[cfg(test)]
pub mod tests {
    use crate::parse2::token::{result, tokenize};
    use crate::parse2::{log, parse_operation};

    #[test]
    pub fn tokenz() {
        let op= parse_operation("tokenz", 
r#"
Release(version=1.3.7) {
  + <SomeClass>;
}       
        "#);
        
       match result(tokenize(op.input())) {
           Ok(_) => {}
           Err(errs) => {
               log(op.data, errs);
           }
       }

        assert_eq!(op.stack.len(), 0)
    }
    
}