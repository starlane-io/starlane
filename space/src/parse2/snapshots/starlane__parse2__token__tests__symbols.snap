---
source: space/src/parse2/token.rs
expression: token
---
Equals
